### PR TITLE
Bug 1971207: AWS CI dynamic availability zone selection

### PR DESCRIPTION
**Pull request by e-tienne**
This PR adds Openshift Installer nodes (Bootstrap, Control and Compute)
instance types availability zones lookup.

https://bugzilla.redhat.com/show_bug.cgi?id=1971207

### CODE REVIEW

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 1]
```diff
@@ -16,10 +16,10 @@ REGION="${LEASED_RESOURCE}"
 
 # Generate working availability zones from the region
 mapfile -t AVAILABILITY_ZONES < <(aws --region "${REGION}" ec2 describe-availability-zones | jq -r '.AvailabilityZones[] | select(.State == "available") | .ZoneName' | sort -u)
-# Generate availability zones with the biggest instance type required
-mapfile -t XLARGE_ZONES < <(aws --region "${REGION}" ec2 describe-instance-type-offerings --location-type availability-zone --filters Name=instance-type,Values=m5.8xlarge | jq -r '.InstanceTypeOfferings[].Location' | sort -u)
+# Generate availability zones with all instance types required
+mapfile -t INSTANCE_ZONES < <(aws --region "${REGION}" ec2 describe-instance-type-offerings --location-type availability-zone --filters Name=instance-type,Values=m5.8xlarge,m5.4xlarge,m5.2xlarge,m5.xlarge,m4.xlarge | jq -r '.InstanceTypeOfferings[].Location' | sort -u)
```
Comment (staebler) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:20]: Would it be easy to get the actual instance types that are going to be used?

Comment (e-tienne) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:20]: That's a good point. I don't see why not.

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 2]
```diff
@@ -14,12 +14,25 @@ function join_by { local IFS="$1"; shift; echo "$*"; }
...
+elif [[ "${SIZE_VARIANT}" == "compact" ]]; then
+  master_type=m5.2xlarge
+fi
+
 # Generate working availability zones from the region
 mapfile -t AVAILABILITY_ZONES < <(aws --region "${REGION}" ec2 describe-availability-zones | jq -r '.AvailabilityZones[] | select(.State == "available") | .ZoneName' | sort -u)
-# Generate availability zones with the biggest instance type required
-mapfile -t XLARGE_ZONES < <(aws --region "${REGION}" ec2 describe-instance-type-offerings --location-type availability-zone --filters Name=instance-type,Values=m5.8xlarge | jq -r '.InstanceTypeOfferings[].Location' | sort -u)
+# Generate availability zones with control and compute instance types required
+mapfile -t INSTANCE_ZONES < <(aws --region "${REGION}" ec2 describe-instance-type-offerings --location-type availability-zone --filters Name=instance-type,Values="${master_type}","${COMPUTE_NODE_TYPE}" | jq -r '.InstanceTypeOfferings[].Location' | sort -u)
```
Comment (staebler) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:33]: We should include the bootstrap type, too, which is currently hardcorded in the installer to `m5.large`.

**Comment by staebler** [comment]
/lgtm
/approve

Jobs are failing due to unrelated API throttling.

Zones are being selected.
```
$ curl https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/19285/rehearse-19285-pull-ci-openshift-origin-release-4.9-e2e-aws/1404873668558852096/artifacts/e2e-aws/ipi-conf-aws/build-log.txt
AWS region: us-west-2 (zones: [ us-west-2a,us-west-2b ])
```

Only zones selected are being used in terraform.
```
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/19285/rehearse-19285-pull-ci-openshift-origin-release-4.9-e2e-aws/1404873668558852096/artifacts/e2e-aws/ipi-install-install/artifacts/.openshift_install.log | grep -E 'Action=CreateSubnet&AvailabilityZone=([^&]+)'
time="2021-06-15T20:40:18Z" level=debug msg="2021-06-15T20:40:18.929Z [DEBUG] plugin.terraform-provider-aws: Action=CreateSubnet&AvailabilityZone=us-west-2b&AvailabilityZoneId=&CidrBlock=10.0.64.0%2F18&Version=2016-11-15&VpcId=vpc-07f5dab6e15dfff3a"
time="2021-06-15T20:40:18Z" level=debug msg="2021-06-15T20:40:18.937Z [DEBUG] plugin.terraform-provider-aws: Action=CreateSubnet&AvailabilityZone=us-west-2a&AvailabilityZoneId=&CidrBlock=10.0.0.0%2F18&Version=2016-11-15&VpcId=vpc-07f5dab6e15dfff3a"
time="2021-06-15T20:40:20Z" level=debug msg="2021-06-15T20:40:20.600Z [DEBUG] plugin.terraform-provider-aws: Action=CreateSubnet&AvailabilityZone=us-west-2a&AvailabilityZoneId=&CidrBlock=10.0.128.0%2F18&Version=2016-11-15&VpcId=vpc-07f5dab6e15dfff3a"
time="2021-06-15T20:40:20Z" level=debug msg="2021-06-15T20:40:20.858Z [DEBUG] plugin.terraform-provider-aws: Action=CreateSubnet&AvailabilityZone=us-west-2b&AvailabilityZoneId=&CidrBlock=10.0.192.0%2F18&Version=2016-11-15&VpcId=vpc-07f5dab6e15dfff3a"
```

**Pull request merged**